src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1250,34 +1317,103 @@ public void OnStartLine(HttpMethod method, HttpVersion version, Span<byte> targe
                 }
             }
 
-            var normalizedTarget = PathNormalizer.RemoveDotSegments(requestUrlPath);
-            if (method != HttpMethod.Custom)
+            QueryString = query.GetAsciiStringNonNullCharacters();
+            RawTarget = rawTarget;
+            SetNormalizedPath(requestUrlPath);
+        }
+
+
+        private void OnAuthorityFormTarget(HttpMethod method, Span<byte> target, Span<byte> line)
+        {
+            // TODO Validate that target is a correct host[:port] string.
+            // Reject as 400 if not. This is just a quick scan for invalid characters
+            // but doesn't check that the target fully matches the URI spec.
+            for (var i = 0; i < target.Length; i++)
```

Comment:
We talked about this the first time I implemented this https://github.com/aspnet/KestrelHttpServer/pull/1246. We didn't end up merging that one because we want to prioritize @pakrym's work on pipelines.